cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,365 @@
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			klog.Warningf("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	f := func(fw godo.Firewall) bool {
+		return fw.Name == fm.workerFirewallName
+	}
+	klog.Info("filtering firewall list for the firewall that has the expected firewall name")
+	fw, err := filterFirewallList(ctx, fm.client, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve list of firewalls from DO API: %v", err)
+	}
+	return fw, nil
+}
+
+// Set applies the given inbound rules to the public access firewall when the current rules and target rules differ.
+// Set will also reconcile away any changes to the outbound rules, name and tags.
+func (fm *firewallManagerOp) Set(ctx context.Context, fr *godo.FirewallRequest) error {
+	targetFirewall := fm.fwCache.getCachedFirewall()
+
+	// A locally cached firewall exists, but the inbound rules don't match the expected
+	// service inbound rules. So we need to use the locally cached firewall ID to attempt
+	// to update the firewall APIs representation of the firewall with the new rules.
+	//
+	// Then we update the local cache with the firewall returned from the Update request.
+	if targetFirewall != nil {
+		if isEqual(targetFirewall, fr) {
+			// A locally cached firewall with matching rules, correct name and tags means there is nothing to update.
+			return nil
+		}
+		currentFirewall, resp, err := fm.client.Firewalls.Update(ctx, targetFirewall.ID, fr)
+		if err != nil {
+			if resp == nil || resp.StatusCode != http.StatusNotFound {
+				return fmt.Errorf("could not update firewall: %v", err)
+			}
+			// Firewall does not exist, so we need to create a new firewall with the
+			// updated inbound rules.
+			currentFirewall, err = fm.createFirewall(ctx, fr.InboundRules)
+			if err != nil {
+				return fmt.Errorf("could not create firewall: %v", err)
+			}
+			klog.Info("successfully created firewall")
+		}
+		fm.fwCache.updateCache(currentFirewall)
+		return nil
+	}
+
+	// Check if the target firewall ID exists but in the case that CCM first starts up and the
+	// firewall ID does not exist yet, we have to retrieve one from the API.
+	if targetFirewall == nil {
+		currentFirewall, err := fm.Get(ctx)
+		if err != nil {
+			return err
+		}
+		if currentFirewall == nil {
+			klog.Info("an existing firewall not found, we need to create one")
+			currentFirewall, err = fm.createFirewall(ctx, fr.InboundRules)
+			if err != nil {
+				return err
+			}
+			klog.Info("successfully created firewall")
+		}
+		fm.fwCache.updateCache(currentFirewall)
+	}
+	return nil
+}
+
+func isEqual(targetFirewall *godo.Firewall, fr *godo.FirewallRequest) bool {
+	equal := true
+	if !cmp.Equal(targetFirewall.Name, fr.Name) {
+		klog.Info("reconciling firewall name")
```

Comment:
I like that we include information about which parts of the firewall were unequal. IMHO though, we shouldn't make these log statements in `isEqual`: the purpose of the method is to compare only; whether we'll reconcile the changes away or not is the responsible of the different part of the logic (namely, `Set()`).

To me, a better approach would be to return a slice of firewall parts that are non-equal, e.g.:

```go
func isEqual(targetFirewall *godo.Firewall, fr *godo.FirewallRequest) (bool, []string) {
    var unequalParts []string
    if targetFirewall.Name != fr.Name {
        unequalParts = append(unequalParts, "name")
    }
    [...]

    return len(unequalParts) == 0, unequalParts
}
```

This would also improve testability of the function.